test/suites/integration/storage_test.go

```diff
@@ -16,104 +16,212 @@ package integration_test
-			}
-		}
-		storageClassName := "ebs-sc-test"
-		bindMode := storagev1.VolumeBindingWaitForFirstConsumer
-		sc := test.StorageClass(test.StorageClassOptions{
-			ObjectMeta: metav1.ObjectMeta{
-				Name: storageClassName,
-			},
-			Provisioner:       aws.String("ebs.csi.aws.com"),
-			VolumeBindingMode: &bindMode,
+var _ = Describe("Persistent Volumes", func() {
+	Context("Static", func() {
+		It("should run a pod with a pre-bound persistent volume (empty storage class)", func() {
+			pvc := test.PersistentVolumeClaim(test.PersistentVolumeClaimOptions{
+				VolumeName:       "test-volume",
+				StorageClassName: lo.ToPtr(""),
+			})
+			pv := test.PersistentVolume(test.PersistentVolumeOptions{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: pvc.Spec.VolumeName,
+				},
+			})
+			pod := test.Pod(test.PodOptions{
+				PersistentVolumeClaims: []string{pvc.Name},
+			})
+
+			env.ExpectCreated(nodeClass, nodePool, pv, pvc, pod)
+			env.EventuallyExpectHealthy(pod)
+			env.ExpectCreatedNodeCount("==", 1)
 		})
-
-		pvc := test.PersistentVolumeClaim(test.PersistentVolumeClaimOptions{
-			ObjectMeta: metav1.ObjectMeta{
-				Name: "ebs-claim",
-			},
-			StorageClassName: aws.String(storageClassName),
-			Resources:        v1.ResourceRequirements{Requests: v1.ResourceList{v1.ResourceStorage: resource.MustParse("5Gi")}},
+		It("should run a pod with a pre-bound persistent volume (non-existent storage class)", func() {
+			pvc := test.PersistentVolumeClaim(test.PersistentVolumeClaimOptions{
+				VolumeName:       "test-volume",
+				StorageClassName: lo.ToPtr("non-existent-storage-class"),
+			})
+			pv := test.PersistentVolume(test.PersistentVolumeOptions{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: pvc.Spec.VolumeName,
+				},
+				StorageClassName: "non-existent-storage-class",
+			})
+			pod := test.Pod(test.PodOptions{
+				PersistentVolumeClaims: []string{pvc.Name},
+			})
+			env.ExpectCreated(nodeClass, nodePool, pv, pvc, pod)
+			env.EventuallyExpectHealthy(pod)
+			env.ExpectCreatedNodeCount("==", 1)
 		})
-
-		pod := test.Pod(test.PodOptions{
-			PersistentVolumeClaims: []string{pvc.Name},
+		It("should run a pod with a pre-bound persistent volume while respecting topology constraints", func() {
+			subnets := env.GetSubnets(map[string]string{"karpenter.sh/discovery": env.ClusterName})
+			shuffledAZs := lo.Shuffle(lo.Keys(subnets))
+
+			pvc := test.PersistentVolumeClaim(test.PersistentVolumeClaimOptions{
+				StorageClassName: lo.ToPtr("non-existent-storage-class"),
+			})
+			pv := test.PersistentVolume(test.PersistentVolumeOptions{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: pvc.Spec.VolumeName,
+				},
+				StorageClassName: "non-existent-storage-class",
+				Zones:            []string{shuffledAZs[0]},
+			})
+			pod := test.Pod(test.PodOptions{
+				PersistentVolumeClaims: []string{pvc.Name},
+			})
+			env.ExpectCreated(nodeClass, nodePool, pv, pvc, pod)
+			env.EventuallyExpectHealthy(pod)
```

Comment:
So in this case, should we unschedule the pod, then create a new pod and expect that pod to be healthy, use the PV, and schedule to the same zone?

Reply:
We could do that but I don't think it's necessary to test the topology logic. The PVC needs to bind to the PV before it can schedule, once it binds the pod gains the PVs topology constraints. IIUC these constraints would cause the pod to fail to schedule if Karpenter launched a node in the wrong AZ.